### PR TITLE
[SPARK-35348][SQL] Support the utils for escapse the regex for ANSI SQL: SIMILAR TO … ESCAPE syntax

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
`ANSI SQL: SIMILAR TO ... ESCAPE` is very useful.
There are some mainstream database support the syntax.
**PostgreSQL**:
https://www.postgresql.org/docs/current/functions-matching.html#FUNCTIONS-SIMILARTO-REGEXP

**Redshift**:
https://docs.aws.amazon.com/redshift/latest/dg/pattern-matching-conditions-similar-to.html

**Sybase**:
http://infocenter.sybase.com/help/index.jsp?topic=/com.sybase.help.sqlanywhere.12.0.0/dbreference/like-regexp-similarto.html

**Firebird**:
http://firebirdsql.org/file/documentation/html/en/refdocs/fblangref25/firebird-25-language-reference.html#fblangref25-commons-predsiimilarto

This util supports the following pattern-matching metacharacters:

Operator | Description
-- | --
% | Matches any sequence of zero or more characters.
_ | Matches any single character.
\| | Denotes alternation (either of two alternatives).
\* | Repeat the previous item zero or more times.
\+ | Repeat the previous item one or more times.
? | Repeat the previous item zero or one time.
{m} | Repeat the previous item exactly m times.
{m,} | Repeat the previous item m or more times.
{m,n} | Repeat the previous item at least m and not more than n times.
() | Parentheses group items into a single logical item.
[...] | A bracket expression specifies a character class, just as in POSIX regular expressions.

**Note**
`SIMILAR TO` is similar to `RLIKE`, but with the following differences:
       1. The `SIMILAR TO` operator returns true only if its pattern matches the entire string,
          unlike `RLIKE` behavior, where the pattern can match any portion of the string.
       2. The regex string allow use _ and % as wildcard characters denoting any single character
          and any string, respectively (these are comparable to . and .* in POSIX regular
          expressions).
       3. The regex string allow use escape character like `LIKE` behavior.
       4. '.', '^' and '$' is not a meta character for `SIMILAR TO`.

### Why are the changes needed?
`ANSI SQL: SIMILAR TO ... ESCAPE` is very useful.

### Does this PR introduce _any_ user-facing change?
Yes, a new feature.

### How was this patch tested?
New tests

### CODE REVIEW

**Comment by beliefer** [comment]
ping @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala [hunk 1]
```diff
@@ -66,6 +67,52 @@ object StringUtils extends Logging {
...
+   *    should convert to quote literal.
+   * 2. If '\' is not escape character, need convert to quote literal.
+   * 3. '.', '^' and '$' is not a meta character for SIMILAR TO,
+   *    so we need convert it to quote literal.
+   * 4. Avoid convert the characters of Java regular expression to quote literal.
+   * @param pattern the SQL pattern to convert
+   * @param escapeChar the escape character.
+   * @return the equivalent Java regular expression of the pattern for SIMILAR TO
+   */
+  def escapeSimilarRegex(pattern: String, escapeChar: Char): String = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:84]: Is it possible to share some code with `escapeLikeRegex`? It seems like the only difference is, let's say `escapeLikeRegex` produces `^the_regex`, and `escapeSimilarRegex` should produce `^the_regex$`

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala:84]: escapeLikeRegex produces the_regex
escapeSimilarRegex produces ^the_regex$

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
ping @cloud-fan

**Comment by beliefer** [comment]
@cloud-fan Please help me to remove stale label.

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
ping @MaxGekk cc @cloud-fan

**Comment by beliefer** [comment]
@cloud-fan Please help me to remove stale label.

**Pull request closed without merging**